### ISSUE TITLE
Reduction of losses: mean vs sum

### ISSUE THREAD

**Issue by bayerj** [blei-lab/edward#311]
Hey,

From what I see, the loss functions in edward are often realized as the _sum_ over data points and dimensions. See e.g. [here](https://github.com/blei-lab/edward/blob/master/edward/inferences/klqp.py#L298).

When tuning hyper parameters, the batch size is often an important one. However, in the case of summing over the data set, it is coupled to the learning rate of a optimizer, as the loss landscape is scaled.

It is therefore good practice to average the data points out, instead of summing over them.

If others think this is a good idea, I can go ahead and do the appropriate changes.

---

PS: When using models that also have spatial dimensions (e.g. `(n_time_steps, n_samples_in_batch, n_dimensons)` for time series data), there should be a convention as well. I prefer to always sum over the last dimension and take the mean over the rest.

**Comment by dustinvtran** [comment]
That's right—preserving the scale for data points is not implemented yet. Converting from sum to mean may not be a solution that works for all settings though. Part of the issue is that only the latent variables associated to a data point (local variables) should be scaled, whereas latent variables shared across data points should not be scaled (global variables). 

I guess what I'm mentioning with hierarchical models is also an example in your PS, where you want to sum over certain dimensions and take the mean of the rest.

An important note is also how to deal with batch training when the user works on a subgraph of the full model (e.g., [`vae.py`](https://github.com/blei-lab/edward/blob/master/examples/vae.py)). With the current API, the user never passes in the full data set size to inference. However, that number is important to appropriately scale gradient computations when updating global variables [1]. (Fortunately, VAEs only use local variables so there's no problem in that example.) So in such a scenario, I think it will be a bit different approach than taking means.

[1] "Derivation of Stochastic Inference in Hierarchical Bayesian Models", http://www.jmlr.org/proceedings/papers/v33/ranganath14-supp.pdf

**Comment by bayerj** [comment]
Yes, you are completely right, I oversaw that. Would it make sense to optionally let the user pass in divisor which could be a tensorflow tensorscalar?

Point is that if you take the minimal example of a vae with single global rv, the corresponding kl term (in case of klqp) will depend on the mini batch size and total number of training points, if you want the loss to be averaged per sample.

**Comment by dustinvtran** [comment]
That's what I think as well, that for certain algorithms we should be able to pass in a scale factor. Most generally, it should one for each latent variable. For example, consider
```python
# p(y1, y2, z1, z2, beta) =
# p(beta)
# [ prod_{n=1}^{N1} p(z1_n | beta) p(y1_n | z1_n, beta) ]
# [ prod_{n=1}^{N2} p(z2_n | beta) p(y2_n | z2_n, beta) ]
# Define a subgraph of the model.
beta = RandomVariable(par=tf.ones(5))
z1 = RandomVariable(par=tf.pack([beta for t in range(M1)]))
z2 = RandomVariable(par=tf.pack([beta for t in range(M2)]))
y1 = RandomVariable(par_1=z par_2=tf.pack([beta for t in range(M1)]))
y2 = RandomVariable(par_1=z par_2=tf.pack([beta for t in range(M2)]))
```
To appropriately scale computation we need a dictionary 
`{z1: float(N1)/M1, z2: float(N2)/M2}`.

Maybe this can be passed in as an initialization argument for certain examples. (I have a short doc page about this that I was working on; I'll clean it up and post it here shortly.)

**Comment by dustinvtran** [comment]
Here's the [doc page](https://cloud.githubusercontent.com/assets/2569867/19622806/4b34b866-9881-11e6-9378-fd7782fa562d.png). It's a WIP and quite long. No need to read all (or any) of it. But the basic idea is that just by having a `scale` argument to `inference.initialize()`, we can support (most?) all our data subsampling desires.
